### PR TITLE
languages: add ROS guide

### PR DESCRIPTION
**Pull request by kyrofa**
My take of the user-story-driven docs for ROS developers.

### CODE REVIEW

**Comment by sergiusens** [comment]
Created the repo @kyrofa https://github.com/snapcraft-docs/ros-talker-listener

File: build-snaps/ros.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
+---
+layout: base
+title: Robot Operating System (ROS)
+---
+
+Snapcraft builds on top of the `catkin` tool, familiar to any ROS developer, to create snaps for people to install on Linux.
```
Comment (sergiusens) [inline build-snaps/ros.md]: this needs some rewording I think, "builds on top of" would be better said as "leverages" I believe.

And "to create snaps for people to install on Linux" has a strange vibe to it, maybe that is just me

File: build-snaps/ros.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+## What problems do snaps solve for ROS applications?
+
+ROS itself is distributed via the OSRF's own Debian archive, along with many community-supported tools. It's possible to get one's own application into their archive as well, but it requires that the application is open-source. One is also left with the question of how to update ROS and one's own application on a robotic platform that has already been shipped. With snapcraft it's just one command to bundle a specific ROS version along with one's own application into a snap that works anywhere and can be automatically updated.
```
Comment (sergiusens) [inline build-snaps/ros.md]: too much use of "one's own" in here.

File: build-snaps/ros.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+Here are some snap advantages that will benefit many ROS projects:
+
+ * Bundle all the runtime requirements, including the exact version of ROS, system libraries, etc.
```
Comment (sergiusens) [inline build-snaps/ros.md]: not sure how welcome this is, but you could potentially have many ROS runtimes in different snaps.

Comment (kyrofa) [reply build-snaps/ros.md]: Yeah, too advanced for this, I think.

File: build-snaps/ros.md (hunk 1 above)
Comment (kyrofa) [reply build-snaps/ros.md]: This is just part of the template.

**Comment by kyrofa** [comment]
> Created the repo @kyrofa https://github.com/snapcraft-docs/ros-talker-listener

Thanks @sergiusens! All set.

File: build-snaps/ros.md (hunk 1 above)
Comment (sergiusens) [reply build-snaps/ros.md]: ok, no change here then, but @evandandrea don't you find this strange after reading?

File: build-snaps/ros.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+confinement: devmode
+```
+
+#### Parts
+
+Parts define how to build your app. Parts can be anything: programs, libraries, or other assets needed to create and run your application. In this case we have one: ros-tutorials. Parts can point to local directories, remote git repositories, or tarballs.
+
+The Catkin plugin will bundle `roscore` in the snap. It will also use `rosdep` to determine the dependencies of the `catkin-packages` provided, download them from the ROS archive, and unpack them into the snap. Finally, it will build the `catkin-packages` specified.
+
+**Important note:** Most ROS developers run out of the `devel` space. As a result, it's easy to forget the importance of good install rules. The Catkin packages you're building must have good install rules, or Snapcraft won't know which components to place into the snap. Make sure you install binaries, libraries, launch files, etc.
```
Comment (sergiusens) [inline build-snaps/ros.md]: you could introduce the `install` scripting stanza to move things over manually though as a poor mans install target.

File: build-snaps/ros.md (hunk 1 above)
Comment (kyrofa) [reply build-snaps/ros.md]: @evandandrea speaks the Queen's English these days, who are we to question? :wink:

File: build-snaps/ros.md (hunk 4 above)
Comment (kyrofa) [reply build-snaps/ros.md]: Good idea. Several of the guides have an "Extending and overriding behaviour" section where they discuss such things. I just added one here.

File: build-snaps/ros.md (hunk 2 above)
Comment (kyrofa) [reply build-snaps/ros.md]: Rephrased to be more personable.

File: build-snaps/ros.md [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+  ros-tutorials:
+    source: https://github.com/ros/ros_tutorials.git
+    source-branch: kinetic-devel
+    plugin: catkin
+    rosdistro: kinetic
+    source-space: roscpp_tutorials/
+    catkin-packages: [roscpp_tutorials]
+    install: |
+      mkdir -p "$SNAPCRAFT_PART_INSTALL/opt/ros/kinetic/share/roscpp_tutorials/launch"
+      cp roscpp_tutorials/launch/talker_listener.launch "$SNAPCRAFT_PART_INSTALL/opt/ros/kinetic/share/roscpp_tutorials/launch/"
```
Comment (sergiusens) [inline build-snaps/ros.md]: only the launch file?

Comment (kyrofa) [reply build-snaps/ros.md]: Made the example more theoretical to avoid confusion.

File: build-snaps/ros.md (hunk 1 above)
Comment (evandandrea) [reply build-snaps/ros.md]: What we were trying to convey with "builds on top of" is that you will continue to work with the tools you're already familiar with. Snapcraft is a light extension on these to produce auto-updating builds for Linux.

If we think this is poorly captured in "builds on top of", that's fine. Let's change it as part of this PR here and in the other language guides. As a general rule, we can and should refine the language used in the guides, making sure to update all of them at once.

Comment (sergiusens) [reply build-snaps/ros.md]: that does make sense, reading it today it does convey that message.

File: build-snaps/ros.md [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+confinement: devmode
+```
+
+#### Parts
+
+Parts define how to build your app. Parts can be anything: programs, libraries, or other assets needed to create and run your application. In this case we have one: ros-tutorials. Parts can point to local directories, remote git repositories, or tarballs.
+
+The Catkin plugin will bundle `roscore` in the snap. It will also use `rosdep` to determine the dependencies of the `catkin-packages` provided, download them from the ROS archive, and unpack them into the snap. Finally, it will build the `catkin-packages` specified.
+
+**Important note:** Most ROS developers run out of the `devel` space. As a result, it's easy to forget the importance of good install rules. The Catkin packages you're building must have good install rules, or Snapcraft won't know which components to place into the snap. Make sure you install binaries, libraries, launch files, etc.
```
Comment (evandandrea) [inline build-snaps/ros.md]: What are "good install rules"?

File: build-snaps/ros.md [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+
+```
+snapcraft help catkin
+```
+
+### Extending and overriding behaviour
+
+You can [extend the behaviour](/build-snaps/scriptlets) of any part in your `snapcraft.yaml` with shell commands. These can be run after pulling the source code but before building by using the `prepare` keyword. The build process can be overridden entirely using the `build` keyword and shell commands. The `install` keyword is used to run shell commands after building your code, useful for making post build modifications such as relocating build assets.
+
+For example, while the ros_tutorials have proper install rules, say you were creating a snap of an upstream ROS application that didn't, and you wanted a launch file out of it. You could make use of the `install` keyword as a poor man's install rule:
```
Comment (evandandrea) [inline build-snaps/ros.md]: I would find a more positive way of describing this use of the install keyword.

File: build-snaps/ros.md (hunk 6 above)
Comment (kyrofa) [reply build-snaps/ros.md]: Install rules that install every component of the package necessary to run. It's fairly common to be missing some or all with ROS packages.

File: build-snaps/ros.md (hunk 7 above)
Comment (kyrofa) [reply build-snaps/ros.md]: I can't think of a more relevant and positive example for ROS. I can remove this section.

Comment (evandandrea) [reply build-snaps/ros.md]: Sorry, let me rephrase. I think the example is fine, but I wouldn’t call it a “poor man’s rule.”

File: build-snaps/ros.md (hunk 6 above)
Comment (evandandrea) [reply build-snaps/ros.md]: Cool. As long as good is unambiguous, I’m +1

**Comment by evandandrea** [comment]
I’m 👍

File: build-snaps/ros.md (hunk 6 above)
Comment (kyrofa) [reply build-snaps/ros.md]: I added a clarification.

File: build-snaps/ros.md (hunk 7 above)
Comment (kyrofa) [reply build-snaps/ros.md]: Ah, I see. Okay, done.

**Comment by kyrofa** [comment]
> I’m :+1:

Your review is still "requesting changes." Mind taking a look when you have a sec?

**Comment by kyrofa** [comment]
(Thanks for the quick turnaround on reviews, by the way. I know you guys have a lot going on.)

**Comment by evandandrea** [comment]
Thanks for this.
